### PR TITLE
[FIRRTL][LowerTypes] Don't silently drop symbols on type-lowered mem's.

### LINKED ISSUE #5592: [FIRRTL][LowerTypes] Memory with symbols blindly drops symbol, spawns new ones?
Code path not under test, should probably be an error (?).

Input:
```
firrtl.circuit "Mem2" {
  firrtl.module @Mem2(in %clock: !firrtl.clock, in %rAddr: !firrtl.uint<4>, in %rEn: !firrtl.uint<1>, out %rData: !firrtl.bundle<a: uint<8>, b: uint<8>>, in %wAddr: !firrtl.uint<4>, in %wEn: !firrtl.uint<1>, in %wMask: !firrtl.bundle<a: uint<1>, b: uint<1>>, in %wData: !firrtl.bundle<a: uint<8>, b: uint<8>>) {
    %memory_r, %memory_w = firrtl.mem sym @X Undefined {depth = 16 : i64, name = "memory", portNames = ["r", "w"], readLatency = 0 : i32, writeLatency = 1 : i32} : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: bundle<a: uint<8>, b: uint<8>>>, !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: uint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>
    %0 = firrtl.subfield %memory_r[clk] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: bundle<a: uint<8>, b: uint<8>>>
    firrtl.connect %0, %clock : !firrtl.clock, !firrtl.clock
    %1 = firrtl.subfield %memory_r[en] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: bundle<a: uint<8>, b: uint<8>>>
    firrtl.connect %1, %rEn : !firrtl.uint<1>, !firrtl.uint<1>
    %2 = firrtl.subfield %memory_r[addr] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: bundle<a: uint<8>, b: uint<8>>>
    firrtl.connect %2, %rAddr : !firrtl.uint<4>, !firrtl.uint<4>
    %3 = firrtl.subfield %memory_r[data] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: bundle<a: uint<8>, b: uint<8>>>
    firrtl.connect %rData, %3 : !firrtl.bundle<a: uint<8>, b: uint<8>>, !firrtl.bundle<a: uint<8>, b: uint<8>>
    %4 = firrtl.subfield %memory_w[clk] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: uint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>
    firrtl.connect %4, %clock : !firrtl.clock, !firrtl.clock
    %5 = firrtl.subfield %memory_w[en] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: uint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>
    firrtl.connect %5, %wEn : !firrtl.uint<1>, !firrtl.uint<1>
    %6 = firrtl.subfield %memory_w[addr] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: uint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>
    firrtl.connect %6, %wAddr : !firrtl.uint<4>, !firrtl.uint<4>
    %7 = firrtl.subfield %memory_w[mask] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: uint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>
    firrtl.connect %7, %wMask : !firrtl.bundle<a: uint<1>, b: uint<1>>, !firrtl.bundle<a: uint<1>, b: uint<1>>
    %8 = firrtl.subfield %memory_w[data] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: uint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>
    firrtl.connect %8, %wData : !firrtl.bundle<a: uint<8>, b: uint<8>>, !firrtl.bundle<a: uint<8>, b: uint<8>>
  }
  sv.verbatim "Testing {{0}}" {symbols = [#hw.innerNameRef<@Mem2::@X>]}
}
```

(Verbatim as generic inner-sym user that would be broken by this behavior)

Running through ` circt-opt --firrtl-lower-types memsym.mlir` produces the following invalid IR (not caught by verified presently but it should):

```
module {
  firrtl.circuit "Mem2" {
    firrtl.module @Mem2(in %clock: !firrtl.clock, in %rAddr: !firrtl.uint<4>, in %rEn: !firrtl.uint<1>, out %rData_a: !firrtl.uint<8>, out %rData_b: !firrtl.uint<8>, in %wAddr: !firrtl.uint<4>, in %wEn: !firrtl.uint<1>, in %wMask_a: !firrtl.uint<1>, in %wMask_b: !firrtl.uint<1>, in %wData_a: !firrtl.uint<8>, in %wData_b: !firrtl.uint<8>) {
      %memory_r_addr = firrtl.wire : !firrtl.uint<4>
      %memory_r_en = firrtl.wire : !firrtl.uint<1>
      %memory_r_clk = firrtl.wire : !firrtl.clock
      %memory_r_data_a = firrtl.wire : !firrtl.uint<8>
      %memory_r_data_b = firrtl.wire : !firrtl.uint<8>
      %memory_w_addr = firrtl.wire : !firrtl.uint<4>
      %memory_w_en = firrtl.wire : !firrtl.uint<1>
      %memory_w_clk = firrtl.wire : !firrtl.clock
      %memory_w_data_a = firrtl.wire : !firrtl.uint<8>
      %memory_w_data_b = firrtl.wire : !firrtl.uint<8>
      %memory_w_mask_a = firrtl.wire : !firrtl.uint<1>
      %memory_w_mask_b = firrtl.wire : !firrtl.uint<1>
      %memory_a_r, %memory_a_w = firrtl.mem sym @Xmemory_a Undefined {depth = 16 : i64, name = "memory_a", portNames = ["r", "w"], readLatency = 0 : i32, writeLatency = 1 : i32} : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: uint<8>>, !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      %memory_b_r, %memory_b_w = firrtl.mem sym @Xmemory_b Undefined {depth = 16 : i64, name = "memory_b", portNames = ["r", "w"], readLatency = 0 : i32, writeLatency = 1 : i32} : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: uint<8>>, !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      %0 = firrtl.subfield %memory_a_r[addr] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: uint<8>>
      firrtl.strictconnect %0, %memory_r_addr : !firrtl.uint<4>
      %1 = firrtl.subfield %memory_b_r[addr] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: uint<8>>
      firrtl.strictconnect %1, %memory_r_addr : !firrtl.uint<4>
      %2 = firrtl.subfield %memory_a_r[en] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: uint<8>>
      firrtl.strictconnect %2, %memory_r_en : !firrtl.uint<1>
      %3 = firrtl.subfield %memory_b_r[en] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: uint<8>>
      firrtl.strictconnect %3, %memory_r_en : !firrtl.uint<1>
      %4 = firrtl.subfield %memory_a_r[clk] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: uint<8>>
      firrtl.strictconnect %4, %memory_r_clk : !firrtl.clock
      %5 = firrtl.subfield %memory_b_r[clk] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: uint<8>>
      firrtl.strictconnect %5, %memory_r_clk : !firrtl.clock
      %6 = firrtl.subfield %memory_a_r[data] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: uint<8>>
      firrtl.strictconnect %memory_r_data_a, %6 : !firrtl.uint<8>
      %7 = firrtl.subfield %memory_b_r[data] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: uint<8>>
      firrtl.strictconnect %memory_r_data_b, %7 : !firrtl.uint<8>
      %8 = firrtl.subfield %memory_a_w[addr] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      firrtl.strictconnect %8, %memory_w_addr : !firrtl.uint<4>
      %9 = firrtl.subfield %memory_b_w[addr] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      firrtl.strictconnect %9, %memory_w_addr : !firrtl.uint<4>
      %10 = firrtl.subfield %memory_a_w[en] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      firrtl.strictconnect %10, %memory_w_en : !firrtl.uint<1>
      %11 = firrtl.subfield %memory_b_w[en] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      firrtl.strictconnect %11, %memory_w_en : !firrtl.uint<1>
      %12 = firrtl.subfield %memory_a_w[clk] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      firrtl.strictconnect %12, %memory_w_clk : !firrtl.clock
      %13 = firrtl.subfield %memory_b_w[clk] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      firrtl.strictconnect %13, %memory_w_clk : !firrtl.clock
      %14 = firrtl.subfield %memory_a_w[data] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      firrtl.strictconnect %14, %memory_w_data_a : !firrtl.uint<8>
      %15 = firrtl.subfield %memory_b_w[data] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      firrtl.strictconnect %15, %memory_w_data_b : !firrtl.uint<8>
      %16 = firrtl.subfield %memory_a_w[mask] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      firrtl.strictconnect %16, %memory_w_mask_a : !firrtl.uint<1>
      %17 = firrtl.subfield %memory_b_w[mask] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: uint<8>, mask: uint<1>>
      firrtl.strictconnect %17, %memory_w_mask_b : !firrtl.uint<1>
      firrtl.connect %memory_r_clk, %clock : !firrtl.clock, !firrtl.clock
      firrtl.connect %memory_r_en, %rEn : !firrtl.uint<1>, !firrtl.uint<1>
      firrtl.connect %memory_r_addr, %rAddr : !firrtl.uint<4>, !firrtl.uint<4>
      firrtl.strictconnect %rData_a, %memory_r_data_a : !firrtl.uint<8>
      firrtl.strictconnect %rData_b, %memory_r_data_b : !firrtl.uint<8>
      firrtl.connect %memory_w_clk, %clock : !firrtl.clock, !firrtl.clock
      firrtl.connect %memory_w_en, %wEn : !firrtl.uint<1>, !firrtl.uint<1>
      firrtl.connect %memory_w_addr, %wAddr : !firrtl.uint<4>, !firrtl.uint<4>
      firrtl.strictconnect %memory_w_mask_a, %wMask_a : !firrtl.uint<1>
      firrtl.strictconnect %memory_w_mask_b, %wMask_b : !firrtl.uint<1>
      firrtl.strictconnect %memory_w_data_a, %wData_a : !firrtl.uint<8>
      firrtl.strictconnect %memory_w_data_b, %wData_b : !firrtl.uint<8>
    }
    sv.verbatim "Testing {{0}}" {symbols = [#hw.innerNameRef<@Mem2::@X>]}
  }
}
```

Code: https://github.com/llvm/circt/blob/db3918a3797ae1a9e066468b63a052a3678a51f1/lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp#L279 .

### PR DESCRIPTION
**Pull request by dtzSiFive**
Fixes #5592.

### CODE REVIEW

File: test/Dialect/FIRRTL/lower-types-issue-5592.mlir [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    %8 = firrtl.subfield %memory_r[clk] : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: vector<uint<8>, 4>>
+    firrtl.strictconnect %8, %clock : !firrtl.clock
+    firrtl.strictconnect %7, %rEn : !firrtl.uint<1>
+    firrtl.strictconnect %6, %rAddr : !firrtl.uint<4>
+    firrtl.strictconnect %rData, %5 : !firrtl.vector<uint<8>, 4>
+    firrtl.strictconnect %4, %clock : !firrtl.clock
+    firrtl.strictconnect %3, %rEn : !firrtl.uint<1>
+    firrtl.strictconnect %2, %rAddr : !firrtl.uint<4>
+    firrtl.strictconnect %1, %wMask : !firrtl.vector<uint<1>, 4>
+    firrtl.strictconnect %0, %wData : !firrtl.vector<uint<8>, 4>
```
Comment (seldridge) [inline test/Dialect/FIRRTL/lower-types-issue-5592.mlir:27]: Is this code relevant to the test?

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp [hunk 2]
```diff
@@ -276,9 +276,9 @@ static MemOp cloneMemWithNewType(ImplicitLocOpBuilder *b, MemOp op,
       op.getRuw(), portNames, (op.getName() + field.suffix).str(),
       op.getNameKind(), op.getAnnotations().getValue(),
       op.getPortAnnotations().getValue(), op.getInnerSymAttr());
-  if (auto oldName = getInnerSymName(op))
-    newMem.setInnerSymAttr(hw::InnerSymAttr::get(StringAttr::get(
-        b->getContext(), oldName.getValue() + (op.getName() + field.suffix))));
+
+  assert(!op.getInnerSym() &&
+         "should already have produced error if sym present");
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:281]: Assert or error and exit?

**Review by seldridge — APPROVED** [review]

File: test/Dialect/FIRRTL/lower-types-issue-5592.mlir (hunk 1 above)
Comment (dtzSiFive) [reply test/Dialect/FIRRTL/lower-types-issue-5592.mlir:27]: No it's not relevant and it's noisy, dropped, thanks!  It makes the IR valid which is why it's included but it's not necessary.

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp (hunk 2 above)
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:281]: See arguments both ways, but since you asked let's make it an error.  It's unreachable presently so no test.
